articles/113_ros_time.md

```diff
@@ -0,0 +1,115 @@
+
+There are however several use cases where being able to control the progress of the system are important.
+
+## Use cases requiring time abstraction
+
+When playing back logged data it is often very valuable to support accelerated, slowed, or stepped control over the progress of time.
+This control can allow you to get to a specific time and pause the system so that you can debug it in depth.
+It is possible to do this with just the playback of the sensor data, however if the sensor data is out of synchronization with the rest of the system it will break many algorithms.
+
+Another important use case for using an abstracted time source is when you are running against a simulated robot instead of a real robot.
+Depending on the simulation characteristics the simulator may be able to run much faster than realtime or much slower.
+Running faster than realtime can be valuable for high level testing as well allowing for repeated system tests.
+Slower than realtime simulation is necessary for complicated systems where accuracy is more important than speed.
+Often the simulation is the limiting factor for the system and as such the simulator can be a time source for faster or slower playback.
+And if the simulation is paused the system can also pause using the same mechanism.
+
+### Challenges in using abstracted time
+
+There are many algorithms for synchronization and they can typically achieve accuracies which are better than the latency of the network communications between devices on the network.
+However, these algorithms take advantage of assumptions about the constant and continuous nature of time.
+
+An important aspect of using an abstracted time is to be able to manipulate time.
+In some cases, changing the speed dynamically for example speeding up or slowing down all the way to pause is important for debugging.
+
+The ability to support pausing time requires that we not assume that the time values are always increasing.
+
+When communicating the changes in time propagation the latencies in the communication network becomes a challenge.
+Any change in the time abstraction must be communicated to the other nodes in the graph, but will be subject to normal network communication latency.
+This inaccuracy is proportional to the latency of communications and also proportional to the increase in the realtime factor.
+As such if very accurate timestamping is required using the time abstraction is required it can be achieved by slowing down the realtime factor such that the communication latency is comparatively small.
+
+The final challenge is that to be useful for log file playback the time abstraction must be able to jump backwards in time.
+This is a very different behavior from most clocks, and requires developers using the time abstraction to make sure their algorithm can deal with the discontinuity.
```

Comment:
The system clock can do this, so maybe instead of saying "This is a very different behavior from most clocks", you could say "This behavior is similar to a system clock after a negative date change".